src/app.rs

```diff
@@ -112,7 +112,26 @@ where
         self
     }
 
-    /// Register a middleware.
+    /// Registers heavyweight Application-level middleware, in the form of a
```

Comment:
i am not sure about `heavyweight` and `lightweight`. should we maybe use extra info for clarification. something like "in form of reusable components", and "in-place request/response modification".
i expect at some point `wrap` and `wrap_fn` will be merged together, but at the moment rust's type inference is not powerful enough for that.